### PR TITLE
Perform checkIODataDirectory() as part of FTS probe.

### PR DESCRIPTION
**Pull request by ashwinstar**

### CODE REVIEW

File: src/backend/fts/ftsmessagehandler.c [hunk 1]
```diff
@@ -19,6 +19,150 @@
...
+ * O_DIRECT flag requires buffer to be OS/FS block aligned.
+ * Best to have it IO Block alligned henece using BLCKSZ
+ */
+static bool
+checkIODataDirectory(void)
+{
+	int fd;
+	int size = BLCKSZ + BLCKSZ;
+	int magic_len = strlen(FTS_PROBE_MAGIC_STRING) + 1;
+	char *data = malloc(size);
```
Comment (danielgustafsson) [inline src/backend/fts/ftsmessagehandler.c]: While not introduced in this PR, why are we using `malloc()` and not `palloc()` here?

File: src/backend/fts/ftsmessagehandler.c [hunk 2]
```diff
@@ -129,6 +273,14 @@ HandleFtsWalRepProbe(void)
...
 	}
 
+	/*
+	 * Perform basic sanity check for disk IO on segment. Without this check
+	 * in many situations FTS didn't detect the problem and hence didn't
+	 * trigger failover to mirror. It caused extended data unavailable
+	 * situations. Hence performing some read-write as part of FTS probe
+	 * helped detect and trigger failover.
+	 */
+	checkIODataDirectory();
```
Comment (asimrp) [inline src/backend/fts/ftsmessagehandler.c]: A new field in `FtsResponse` should be added to indicate if disk I/O has failed.  And FTS on master should consider the primary that returns such a response as down.

**Comment by ashwinstar** [comment]
On Fri, Jan 19, 2018 at 12:18 PM, Asim RP <notifications@github.com> wrote:

> *@asimrp* commented on this pull request.
> ------------------------------
>
> In src/backend/fts/ftsmessagehandler.c
> <https://github.com/greenplum-db/gpdb/pull/4366#discussion_r162722573>:
>
> > @@ -129,6 +273,14 @@ HandleFtsWalRepProbe(void)
>  		elog(LOG, "received probe message while acting as mirror");
>  	}
>
> +	/*
> +	 * Perform basic sanity check for disk IO on segment. Without this check
> +	 * in many situations FTS didn't detect the problem and hence didn't
> +	 * trigger failover to mirror. It caused extended data unavailable
> +	 * situations. Hence performing some read-write as part of FTS probe
> +	 * helped detect and trigger failover.
> +	 */
> +	checkIODataDirectory();
>
> A new field in FtsResponse should be added to indicate if disk I/O has
> failed. And FTS on master should consider the primary that returns such a
> response as down.
>
Agreed :-) just was trying to fix the warning soonish.

**Comment by asimrp** [comment]
> Agreed :-) just was trying to fix the warning soonish.

In that case, how about `elog(ERROR)` from `checkIODataDirectory()` instead of returning `false`?  FTS should see that as a probe failure.

**Comment by ashwinstar** [comment]
On Fri, Jan 19, 2018 at 12:36 PM, Asim RP <notifications@github.com> wrote:

> Agreed :-) just was trying to fix the warning soonish.
>
> In that case, how about elog(ERROR) from checkIODataDirectory() instead
> of returning false? FTS should see that as a probe failure.
>
Will hold the PR for a day or two and code the final thing we wish to code
anyways.

File: src/backend/fts/ftsmessagehandler.c [hunk 3]
```diff
@@ -19,6 +19,150 @@
...
+				(errcode_for_file_access(),
+				errmsg("could not unlink file \"%s\": %m", FTS_PROBE_FILE_NAME)));
+		}
+	}
+
+	if(data)
+		free(data);
+
+	return failure;
+}
```
Comment (ivannovick) [inline src/backend/fts/ftsmessagehandler.c]: nice IO check, cheers

**Comment by ashwinstar** [comment]
On Fri, Jan 19, 2018 at 12:36 PM, Asim RP <notifications@github.com> wrote:

> Agreed :-) just was trying to fix the warning soonish.
>
> In that case, how about elog(ERROR) from checkIODataDirectory() instead
> of returning false? FTS should see that as a probe failure.
>
Thinking more seems yes that good for current scope. Adding extra boolean
just for this case where action would be same for it on other end to make
primary down doesn't;t seem much justified. So, made change to elog(ERROR).

File: src/backend/fts/ftsmessagehandler.c (hunk 1 above)
Comment (ashwinstar) [reply src/backend/fts/ftsmessagehandler.c]: I think earlier it was invoked by a process which was not really a backend and didn't have all the infra initialized to perform palloc. Made it palloc0 now as now FTS message handler is backend process.

**Comment by ashwinstar** [comment]
On Fri, Jan 26, 2018 at 10:12 AM, Asim RP <notifications@github.com> wrote:

> *@asimrp* approved this pull request.
>
> One caveat - if the disk is full, writing the error message to log file
> may fail. Will the error make it back to FTS on master in this case?
>
Yes, plus even if it doesn't it's same as IO hang, FTS retries and finally
marks primary down.

**Pull request merged**